modules/transport/xmlrpc/main.c

```diff
@@ -511,49 +514,63 @@ xmlrpcmethod_metadata(void *conn, int parc, char *parv[])
 		}
 	}
 
-	if (parc < 2)
+	if (parc < 4)
 	{
 		xmlrpc_generic_error(fault_needmoreparams, "Insufficient parameters.");
 		return 0;
 	}
 
-	if (*parv[0] == '#')
+	if ((mu = myuser_find(parv[1])) == NULL)
+	{
+		xmlrpc_generic_error(fault_nosuch_source, "Unknown user.");
+		return 0;
+	}
+
+	if (authcookie_validate(parv[0], mu) == false)
+	{
+		xmlrpc_generic_error(fault_badauthcookie, "Invalid authcookie for this account.");
+		return 0;
+	}
+
+	if (*parv[2] == '#')
 	{
 		struct mychan *mc;
 
-		mc = mychan_find(parv[0]);
+		mc = mychan_find(parv[2]);
 		if (mc == NULL)
 		{
 			xmlrpc_generic_error(fault_nosuch_source, "No channel registration was found for the provided channel name.");
 			return 0;
 		}
 
-		md = metadata_find(mc, parv[1]);
+		md = metadata_find(mc, parv[3]);
 	}
 	else
 	{
 		struct myentity *mt;
 
-		mt = myentity_find(parv[0]);
+		mt = myentity_find(parv[2]);
 		if (mt == NULL)
-			mt = myentity_find_uid(parv[0]);
+			mt = myentity_find_uid(parv[2]);
 
 		if (mt == NULL)
 		{
 			xmlrpc_generic_error(fault_nosuch_source, "No account was found for this accountname or UID.");
 			return 0;
 		}
 
-		md = metadata_find(mt, parv[1]);
+		md = metadata_find(mt, parv[3]);
 	}
 
 	if (md == NULL)
 	{
 		xmlrpc_generic_error(fault_nosuch_source, "No metadata found matching this account/channel and key.");
 		return 0;
 	}
-
-	xmlrpc_string(buf, md->value);
+	if (!strncmp(md->name, "private:", 8) && is_soper(mu))
+		xmlrpc_string(buf, md->value);
+	else
+		xmlrpc_string(buf, "");
```

Comment:
Just one more thing that I missed last time.

Upon closer inspection, this has a bug; it no longer shows any metadata values unless they're private metadata and the user is a soper. Also, just checking whether the user is a soper is not granular enough; many utility bots are given soper status with only the privileges they need to perform their function (e.g. `operserv:akill`). We have specific `*:auspex` privileges for viewing private information about channels, groups, and users, so it makes sense that this should check for those privileges too.

Therefore, this block needs changing to

```C
bool showprivate = false;

if (*parv[2] == '#' && has_priv_myuser(mu, PRIV_CHAN_AUSPEX))
    showprivate = true;
else if (*parv[2] == '!' && has_priv_myuser(mu, PRIV_GROUP_AUSPEX))
    showprivate = true;
else if (*parv[2] != '#' && *parv[2] != '!' && has_priv_myuser(mu, PRIV_USER_AUSPEX))
    showprivate = true;

if (strncmp(md->name, "private:", 8) != 0 || showprivate)
    xmlrpc_string(buf, md->value);
else
    xmlrpc_string(buf, "");
```